prov/util/src/util_mr_cache.c

```diff
@@ -0,0 +1,1139 @@
+					" Have you freed memory w/o closing the memory region?\n");
+				util_mr_cache_notifier_warned = 1;
+			}
+
+			FI_DBG(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+			       "Marking unmapped entry (%p) as retired %"PRIu64":%"PRIu64"\n",
+			       entry, entry->region.address, entry->region.length);
+
+			util_mr_cache_entry_set_unmapped(entry);
+
+			if (util_mr_cache_entry_is_retired(entry))
+				/* Nothing to do */
+				break;
+
+			/* Retire this entry (remove from inuse tree) */
+			util_mr_cache_entry_set_retired(entry);
+			iter = rbtFind(cache->inuse_tree, &entry->region);
+			if (OFI_LIKELY(iter != NULL)) {
+				ret = rbtErase(cache->inuse_tree, iter);
+				if (OFI_UNLIKELY(ret != RBT_STATUS_OK)) {
+					FI_WARN(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+						"Unmapped entry could not be removed from in usetree.\n");
+					abort();
+				}
+			} else {
+				/*  The only way we should get here is if we're in the
+				 *  middle of retiring this entry.  Not sure if this
+				 *  is worth a separate warning from the one above.
+				 */
+			}
+
+			break;
+		} else {
+			util_mr_cache_entry_set_unmapped(entry);
+			iter = rbtFind(cache->stale_tree, &entry->region);
+			if (!iter)
+				break;
+
+			ret = rbtErase(cache->stale_tree, iter);
+			if (OFI_UNLIKELY(ret != RBT_STATUS_OK)) {
+				FI_WARN(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+					"Unmapped entry could not be removed from stale tree.\n");
+				abort();
+			}
+
+			FI_DBG(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+			       "Removed unmapped entry (%p) from stale tree %"PRIu64":%"PRIu64"\n",
+			       entry, entry->region.address, entry->region.length);
+
+			util_mr_cache_lru_remove(cache, entry);
+			FI_DBG(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+			       "Removed unmapped entry (%p) from lru list %"PRIu64":%"PRIu64"\n",
+			       entry, entry->region.address, entry->region.length);
+			cache->stale_elem--;
+
+			util_mr_cache_entry_destroy(cache, entry);
+
+			break;
+		}
+	}
+	if (OFI_UNLIKELY(ret != -FI_EAGAIN))
+		/* Should we do something else here? */
+		FI_WARN(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+			"`get event` returned error: %s\n",
+			fi_strerror(-ret));
+
+	return;
+}
+
+static int
+util_mr_cache_notifier_monitor(struct ofi_mr_cache *cache,
+			       struct ofi_mr_cache_entry *entry)
+{
+
+    if (!UTIL_MR_CACHE_IS_LAZY_DEREG(cache))
+		return FI_SUCCESS;
+
+	FI_DBG(UTIL_MR_CACHE_GET_PROV(cache), FI_LOG_MR,
+	       "monitoring entry=%p %"PRIu64":%"PRIu64"\n",
+	       entry, entry->region.address, entry->region.length);
+
+	return ofi_monitor_subscribe(&cache->mem_nq,
+				     (void *)entry->region.address,
+				     entry->region.length,
+				     &entry->subscription);
+}
+
+static void
+util_mr_cache_notifier_unmonitor(struct ofi_mr_cache *cache,
+				 struct ofi_mr_cache_entry *entry)
+{
+
+	if (!UTIL_MR_CACHE_IS_LAZY_DEREG(cache))
+		return;
+
+	util_mr_cache_clear_notifier_events(cache);
+
+	if (!util_mr_cache_entry_is_unmapped(entry)) {
```

Comment:
This may not make sense given the implementation of the ofi_monitor.

Reply:
Sorry, but I'm not sure that I got it :) Did you mean that `util_mr_cache_entry_is_unmapped` is useless here or?
Thanks in advance for the explanation!

Reply:
So, the KDREG monitoring system worked in the manner that if a unmapped event was received, it was automatically unmonitored. In the implementation of the ofi_monitor, each region needs to be explicitly unmonitored. 

What I'm getting at here is that the unmonitor action should be done unconditionally here, rather than only when the entry is still mapped.